src/libraries/Common/src/Interop/OSX/Interop.NetworkTls.cs

```diff
@@ -0,0 +1,153 @@
+                                                               ref void* negotiatedAlpn, out uint alpnLength);
+
+            [LibraryImport(Interop.Libraries.AppleNetworkNative, EntryPoint = "AppleNetNative_NwCopyCertChain")]
+            internal static partial int CopyCertChain(SafeNetworkFrameworkHandle connection,
+                                                             out SafeCFArrayHandle certificates,
+                                                             out int count);
+
+            internal static int GetAlpnProtocolListSerializedLength(List<SslApplicationProtocol>? applicationProtocols)
+            {
+                if (applicationProtocols is null)
+                {
+                    return 0;
+                }
+
+                int protocolSize = 0;
+
+                foreach (SslApplicationProtocol protocol in applicationProtocols)
+                {
+                    if (protocol.Protocol.Length == 0 || protocol.Protocol.Length > byte.MaxValue)
+                    {
+                        throw new ArgumentException(SR.net_ssl_app_protocols_invalid, nameof(applicationProtocols));
+                    }
+
+                    protocolSize += protocol.Protocol.Length + 2;
+                }
+
+                return protocolSize;
+            }
+
+            private static void SerializeAlpnProtocolList(List<SslApplicationProtocol>? applicationProtocols, Span<byte> buffer)
+            {
+                if (applicationProtocols is null)
+                {
+                    return;
+                }
+
+                Debug.Assert(GetAlpnProtocolListSerializedLength(applicationProtocols) == buffer.Length);
+
+                int offset = 0;
+                foreach (SslApplicationProtocol protocol in applicationProtocols)
+                {
+                    buffer[offset++] = (byte)protocol.Protocol.Length;
+                    protocol.Protocol.Span.CopyTo(buffer.Slice(offset));
+                    offset += protocol.Protocol.Length;
+                    buffer[offset++] = 0;
+                }
+            }
+        }
+        // Status enumeration for Network Framework TLS operations
+        internal enum StatusUpdates
+        {
+            UnknownError = 0,
+            FramerStart = 1,
+            FramerStop = 2,
+            HandshakeFinished = 3,
+            HandshakeFailed = 4,
+            ConnectionReadFinished = 100,
+            ConnectionWriteFinished = 101,
+            ConnectionWriteFailed = 102,
+            ConnectionCancelled = 103,
+        }
+
+        internal enum OSStatus
+        {
+            NoError = 0,
+            ReadError = -19,
+            WriteError = -20,
+            EOFError = -39,
+            SecUserCanceled = -128,
+            WouldBlock = -9803
+        }
+    }
+
+    // Safe handle classes for Network Framework TLS resources
+    internal sealed class SafeNetworkFrameworkHandle : SafeHandleZeroOrMinusOneIsInvalid
```

Comment:
`Nw` abbreviation seems to be used a lot in the platfrom API, I would consider naming the safe handle `SafeNwHandle`